### PR TITLE
Implement the logic of checking node's local PValus using FisherZ test then compare the result with Uniform Distribution using Anderson Darling test

### PR DESCRIPTION
**Pull request by vbcwonderland**
Implement the logic of checking node's local PValus using FisherZ test then compare the result with Uniform Distribution using Anderson Darling test. 

Still need to implement the logic of connecting this logic with the UI part.

### CODE REVIEW

File: tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java [hunk 1]
```diff
@@ -635,13 +673,13 @@ public String getMarkovCheckRecordString() {
...
-               "Number of tests (indep): " + record.numTestsInd + "\n" +
-               "Number of tests (dep): " + record.numTestsDep;
+                "Anderson-Darling p-value (dep): " + nf.format(record.adDep) + "\n" +
+                "Binomial p-value (indep): " + nf.format(record.binIndep) + "\n" +
+                "Binomial p-value (dep): " + nf.format(record.binDep) + "\n" +
+                "Fraction of dependent judgments (indep): " + nf.format(record.fracDepInd) + "\n" +
+                "Fraction of dependent judgments (dep): " + nf.format(record.fracDepDep) + "\n" +
+                "Number of tests (indep): " + record.numTestsInd + "\n" +
+                "Number of tests (dep): " + record.numTestsDep;
     }
```
Comment (vbcwonderland) [inline tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java:683]: Hi Joe, I will change back this bit, as this change is done by auto formatter.

File: tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java [hunk 2]
```diff
@@ -560,9 +598,9 @@ public Knowledge getKnowledge() {
     public void setKnowledge(Knowledge knowledge) {
         if (!(knowledge.getListOfExplicitlyForbiddenEdges().isEmpty() && knowledge.getListOfRequiredEdges().isEmpty())) {
             throw new IllegalArgumentException("Knowledge object for the Markov checker cannot contain required of " +
-                                               "explicitly forbidden edges; only tier knowledge is used. The last tier contains the possible X " +
-                                               "and Y for X _||_ Y | Z1,..,Zn, and the previous tiers contain the possible Z1,..,Zn for X _||_ Y " +
-                                               "| Z1,..,Zn.");
+                    "explicitly forbidden edges; only tier knowledge is used. The last tier contains the possible X " +
+                    "and Y for X _||_ Y | Z1,..,Zn, and the previous tiers contain the possible Z1,..,Zn for X _||_ Y " +
+                    "| Z1,..,Zn.");
```
Comment (vbcwonderland) [inline tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java:603]: Hi Joe, I will change back this bit, as this change is done by auto formatter.

File: tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java [hunk 3]
```diff
@@ -793,7 +831,7 @@ private void addResults(Set<IndependenceResult> resultsIndep, Set<IndependenceRe
                     throw new RuntimeException(e);
                 }
                 boolean indep = result.isIndependent();
-                double pValue = result.getPValue(); // VBC TODO: is this the real pvalue for node x? or is it just storing d-speration fact? in IndTestFisherZ.java this is a real p value for x, given y, z.
+                double pValue = result.getPValue();
```
Comment (vbcwonderland) [inline tetrad-lib/src/main/java/edu/cmu/tetrad/search/MarkovCheck.java:834]: Ignore this, cleanup previous helper comment.

**Comment by jdramsey** [comment]
OK! Sorry about the delay; my head was in a bunch of code. Let me look at this now.

**Review by jdramsey — COMMENTED** [review]
Oh I see--here:

        List<IndependenceFact> factList = new ArrayList<>();
        for (Node y : graphNodes_others) {
            // TODO: shall i check if independenceTest is msepTest?
            IndependenceResult testRes = independenceTest.checkIndependence(x, y, parents);
            if (testRes.isValid()) factList.add(testRes.getFact());
        }

You can make a new MsepTest based on the true graph, like so:

MsepTest msepTest = new MsepTest(graph);

Below, you shouldn't need to check that the test is Fisher Z; someone else in the future might want to use a different test like ChiSquare, that can produce a p-value.

**Comment by vbcwonderland** [comment]
Hi Joe, 
I created the MsepTest instance.  : ) 
The Fisher Z check part below is an issue though. I think right now we have to do it as: 
```
List<Double> pVals = new ArrayList<>();
        for (IndependenceFact f : facts) {
            double pV;
            // For now, check if the test is FisherZ test.
            if (independenceTest instanceof IndTestFisherZ) {
                pV = ((IndTestFisherZ)independenceTest).getPValue(f.getX(), f.getY(), f.getZ());
                pVals.add(pV);
            }
        }
```
Because `getPValue` is not within IndependenceTest public interface. In case of ChiSquare, `getPVal()` has a different function signature, and it also not directly on `IndTestChiSquare` but on its result class in `ChiSquareTest`class 's `ChiSquareTest.Result`. Usage example: 

(in `IndTestChiSquare.java`)
```
...
private final Map<IndependenceFact, ChiSquareTest.Result> facts = new ConcurrentHashMap<>();
...
ChiSquareTest.Result result = this.facts.get(new IndependenceFact(x, y, _z));
            return new IndependenceResult(new IndependenceFact(x, y, _z), result.isIndep(), result.getPValue(),
                    getAlpha() - result.getPValue());
```
Can we add a `double getPValue(Node x, Node y, Set<Node> z)` public method in `IndTestChiSquare`? 
so that we can access its `private final Map<IndependenceFact, ChiSquareTest.Result> facts = new ConcurrentHashMap<>();` by creating a `IndependenceFact` usng x, y, z and return the corresponding `ChiSquareTest.Result` to further call the result class's pvalue?

**Comment by jdramsey** [comment]
That's a good idea. Do you want to do it, or should I?

**Comment by vbcwonderland** [comment]
I can do it:D should be quick. maybe on a seperate Pull Request?

**Comment by jdramsey** [comment]
Sure! Bryan Andrews agreed to review my pull request, in case you were wondering. Hopefully by tonight. Nothing should conflict with you.

**Comment by vbcwonderland** [comment]
@jdramsey the other PR is ready:D https://github.com/cmu-phil/tetrad/pull/1755

**Comment by jdramsey** [comment]
One second, let me look...

**Review by jdramsey — APPROVED** [review]
Nice!

**Pull request merged**